src/librustc/ty/layout.rs

```diff
@@ -1621,25 +1621,11 @@ impl<'a, 'tcx, C> TyLayoutMethods<'tcx, C> for Ty<'tcx>
                 match tcx.struct_tail(pointee).sty {
                     ty::TySlice(_) |
                     ty::TyStr => tcx.types.usize,
-                    ty::TyDynamic(data, _) => {
-                        let trait_def_id = data.principal().unwrap().def_id();
-                        let num_fns: u64 = crate::traits::supertrait_def_ids(tcx, trait_def_id)
-                            .map(|trait_def_id| {
-                                tcx.associated_items(trait_def_id)
-                                    .filter(|item| item.kind == ty::AssociatedKind::Method)
-                                    .count() as u64
-                            })
-                            .sum();
+                    ty::TyDynamic(_, _) => {
                         tcx.mk_imm_ref(
                             tcx.types.re_static,
-                            tcx.mk_array(tcx.types.usize, 3 + num_fns),
-                        )
-                        /* FIXME use actual fn pointers
-                        tcx.mk_tup(&[
                             tcx.mk_array(tcx.types.usize, 3),
-                            tcx.mk_array(Option<fn()>),
-                        ])
-                        */
+                        )
```

Comment:
Huh, this is a *bit* surprising to me, but I'm gonna trust you all here =)

Reply:
So...the issue is that functions with where bounds also contributed to the count, even if they were excluded because they aren't object safe and thus they shouldn't be in the vtable. the previous code was simply wrong, and fixing it fully would blow anything we could beta backport

Reply:
The same patch should also land on master though as the layout is still wrong there?